fontbe/src/features/kern.rs

```diff
@@ -359,31 +382,719 @@ impl Work<Context, AnyWorkId, Error> for KerningGatherWork {
+            (x.into(), y.into())
+        }
+
+        let marks = self.mark_glyph_ids();
+        if marks.is_empty() {
+            return (
+                pairs.iter().map(|x| Cow::Borrowed(*x)).collect(),
+                Vec::new(),
+            );
+        }
+
+        let (mut base_pairs, mut mark_pairs) = (Vec::new(), Vec::new());
+
+        for pair in pairs {
+            match pair {
+                PairPosEntry::Pair(side1, _, side2, _)
+                    if !marks.contains(side1) && !marks.contains(side2) =>
+                {
+                    base_pairs.push(Cow::Borrowed(*pair))
+                }
+                PairPosEntry::Pair(..) => mark_pairs.push(Cow::Borrowed(*pair)),
+
+                // handle the case where all are marks or bases, first:
+                PairPosEntry::Class(side1, val1, side2, val2) => {
+                    let side1_cls = classify_glyphset_contents(side1, &marks);
+                    let side2_cls = classify_glyphset_contents(side2, &marks);
+
+                    match (side1_cls, side2_cls) {
+                        (GlyphSetContent::Empty, _) | (_, GlyphSetContent::Empty) => continue,
+                        (GlyphSetContent::BasesOnly, GlyphSetContent::BasesOnly) => {
+                            base_pairs.push(Cow::Borrowed(*pair));
+                            continue;
+                        }
+                        (GlyphSetContent::BasesOnly, GlyphSetContent::MarksOnly)
+                        | (GlyphSetContent::MarksOnly, GlyphSetContent::BasesOnly)
+                        | (GlyphSetContent::MarksOnly, GlyphSetContent::MarksOnly) => {
+                            mark_pairs.push(Cow::Borrowed(*pair));
+                            continue;
+                        }
+                        _ => {
+                            let (side1_bases, side1_marks) = split_glyphs(side1, &marks);
+                            let (side2_bases, side2_marks) = split_glyphs(side2, &marks);
+
+                            if !side1_bases.is_empty() && !side2_bases.is_empty() {
+                                base_pairs.push(Cow::Owned(PairPosEntry::Class(
+                                    side1_bases.clone(),
+                                    val1.clone(),
+                                    side2_bases.clone(),
+                                    val2.clone(),
+                                )));
+                            }
+                            // these various combos all go in the marks group
+                            for (side1, side2) in [
+                                (&side1_bases, &side2_marks),
+                                (&side1_marks, &side2_bases),
+                                (&side1_marks, &side2_marks),
+                            ] {
+                                if !side1.is_empty() && !side2.is_empty() {
+                                    base_pairs.push(Cow::Owned(PairPosEntry::Class(
```

Comment:
@cmyr If I followed the logic correctly, shouldn't this `base_pairs` actually be `mark_pairs`?

Reply:
good catch, I have tests for this in #733 but apparently not handling the class case, will fix that now :)